File: CS Course/5 Design Patterns/3 Behavioral Pattern/1_Command/Command.py
'''
Author: [lxp]
Date: 2022-01-20 16:04:16
LastEditors: [lxp]
LastEditTime: 2022-01-20 16:20:52
Description: 
命令模式(Command Pattern)：将一个请求封装为一个对象，从而使我们可用不同的请求对客户进行参数化；对请求排队或者记录请求日志，以及支持可撤销的操作。命令模式是一种对象行为型模式，其别名为动作(Action)模式或事务(Transaction)模式。
'''
class Receiver(object):

    def action(self):
        print("Receiver::action")


class Command(object):

    def execute(self):
        raise NotImplementedError


class ComcreteCommand(Command):

    def __init__(self, receiver):
        self._receiver = receiver
    
    def execute(self):
        print("Executing ComcreteCommand")
        self._receiver.action()


class Invoker(object):

    def __init__(self, cmd = None):
        self._cmd = cmd

    def call(self):
        print("Calling Invoker")
        if self._cmd is not None:
            self._cmd.execute()


if __name__ == '__main__':
    rcv = Receiver()
    cmd = ComcreteCommand(rcv)
    ivk = Invoker(cmd)
    ivk.call()

File: CS Course/5 Design Patterns/3 Behavioral Pattern/1_Command/Command.cpp
/*
 * @Author: [lxp]
 * @Date: 2022-01-20 16:04:21
 * @LastEditors: [lxp]
 * @LastEditTime: 2022-01-20 16:12:34
 * @Description: 
 */
#include <iostream>


class Receiver
{
public:
  void action()
  {
    std::cout << "Receiver: execute action" << std::endl;
  }
  // ...
};


class Command
{
public:
  virtual ~Command() {}
  virtual void execute() = 0;
  // ...

protected:
  Command() {}
};


class ConcreteCommand : public Command
{
public:
  ConcreteCommand( Receiver *r ) : receiver( r ) {}
  
  ~ConcreteCommand()
  {
    if ( receiver )
    {
      delete receiver;
    }
  }
  
  void execute()
  {
    receiver->action();
  }
  // ...
  
private:
  Receiver *receiver;
  // ...
};


class Invoker
{
public:
  void set( Command *c )
  {
    command = c;
  }
  
  void confirm()
  {
    if ( command )
    {
      command->execute();  
    }
  }
  // ...

private:
  Command *command;
  // ...
};


int main()
{
  ConcreteCommand command( new Receiver() );
  
  Invoker invoker;
  invoker.set( &command );
  invoker.confirm();
  
  return 0;
}